File: Fortran/IterateBellman.f90
SUBROUTINE IterateBellman

USE Parameters
USE Globals
USE Procedures
USE mUMFPACK

IMPLICIT NONE

INTEGER			:: iaby,ii
REAL(8)			:: lVdiff,ltau,ltau0,ladrift,lc,lh,llabdisutil

delta = deltass
	
!set drifts
ltau = 15.0
ltau0 = (ra+PerfectAnnuityMarkets*deathrate)*(agrid(ngpa)*0.999)**(1.0-ltau)
adrift = ra*agrid + PerfectAnnuityMarkets*deathrate*agrid - ltau0*(agrid**ltau)
bdrift = MERGE((rb+PerfectAnnuityMarkets*deathrate)*bgrid,(rborr+PerfectAnnuityMarkets*deathrate)*bgrid,bgrid>0.0)

IF (Borrowing==1 .and. bgrid(1) < -lumptransfer/(rborr+PerfectAnnuityMarkets*deathrate)) THEN
	write(*,*) 'Warning: natural borrowing limit violated'
END IF


!Initial Guess
IF	(	(EquilibriumR==0 .and. calibrating==.false.) &
 	.or. (EquilibriumR==1 .and. neqmiter<=3 .and. calibrating==.false.) &
 	.or. (CalibrateDiscountRate==1 .and. neqmiter<=3 .and. calibrating==.false.) &
 	.or. (CalibrateRhoAtInitialGuess==1 .and. neqmiter<=3 .and. calibrating==.false.) &
	.or. (calibrating==.true. .and. ImposeEqumInCalibration==1  .and. neqmiter==1 ) &
	.or. (calibrating==.true. .and. ImposeEqumInCalibration==0 ))  THEN
	
	!$OMP PARALLEL DO PRIVATE (ladrift,lh,lc,llabdisutil)
	DO iaby = 1,naby
		IF(LaborSupplyGHH==1) 	lh = (netwage*ygrid(yfromaby(iaby)))**frisch
		IF(LaborSupplySep==1) 	lh = 1.0/3.0
		IF(NoLaborSupply==1) 	lh = 1.0
  		ladrift = (ra + PerfectAnnuityMarkets*deathrate)*agrid(afromaby(iaby))
		lc = netwage*ygrid(yfromaby(iaby))*lh + lumptransfer + (rb+PerfectAnnuityMarkets*deathrate)*bgrid(bfromaby(iaby))
		llabdisutil = lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)	
		IF(NoLaborSupply==1 .or. LaborSupplyGHH==1)V(afromaby(iaby),bfromaby(iaby),yfromaby(iaby)) = utilfn(lc)  / (rho+deathrate)
		IF(LaborSupplySep==1)V(afromaby(iaby),bfromaby(iaby),yfromaby(iaby)) = (utilfn(lc) - chi*llabdisutil) / (rho+deathrate)
	END DO
	!$OMP END PARALLEL DO
	
END IF



ii = 1
lVdiff = 1.0
DO WHILE (ii<=maxiter .and. lVdiff>Vtol)

	CALL HJBUpdate

	!check for convergence
	lVdiff = maxval(abs(Vnew-V))
	IF(Display>=2) write(*,*) "Iteration: ",ii," max V change: ",lVdiff
	
	!update
	V = Vnew
	ii = ii+1
	
END DO



END SUBROUTINE IterateBellman

File: mycode/IterateBellman.m
delta = deltass;
	
% set drifts
ltau = 15;
ltau0 = (ra+PerfectAnnuityMarkets*deathrate)*(amax*0.999)^(1-ltau);
adrift = (ra+PerfectAnnuityMarkets*deathrate)*agrid - ltau0*agrid.^ltau;
bdrift = (rborr*(bgrid<=0)+rb*(bgrid>0)+PerfectAnnuityMarkets*deathrate).*bgrid;

if Borrowing && bgrid(1) < -lumptransfer/(rborr+PerfectAnnuityMarkets*deathrate)
	fprintf('Warning: natural borrowing limit violated\n');
end

% Initial Guess
if (~EquilibriumR && ~calibrating) || ...
    (EquilibriumR && neqmiter<=3 && ~calibrating) || ...
    (CalibrateDiscountRate && neqmiter<=3 && ~calibrating) || ...
    (CalibrateRhoAtInitialGuess && neqmiter<=3 && ~calibrating) || ...
    (calibrating && ImposeEqumInCalibration && neqmiter==1 ) || ...
    (calibrating && ~ImposeEqumInCalibration)
	
    switch LaborSupply
        case 0; lh = 1;
        case 1; lh = 1/3;
        case 2; lh = (netwage*ygrid).^frisch;
    end
%     ladrift = (ra + PerfectAnnuityMarkets*deathrate)*agrid;
    lc = (netwage*lh).*ygrid + lumptransfer + (rb+PerfectAnnuityMarkets*deathrate)*bgrid;
    V = utilfn(lc)/(rho+deathrate);
    if LaborSupply==1
        V = V - chi/(rho+deathrate)/(1+1/frisch) * lh.^(1+1/frisch);
    end
    V = repmat(V,ngpa,1,1);
end
ii = 1;
lVdiff = 1.0;
while ii<=maxiter && lVdiff>Vtol
	[Vnew,c,h,s,d,dadj] = HJBUpdate(V,adrift,bdrift,netwage,lumptransfer,labtax,profit,meanlabeff,delta);
	% check for convergence
	lVdiff = max(abs(Vnew(:)-V(:)));
	if Display>1; fprintf('Iteration: %3d    max V change: %.15g\n',ii,lVdiff); end

	% update
	V = Vnew;
	ii = ii+1;
end